lib/rubocop/cop/style/legacy_hash.rb

```diff
@@ -0,0 +1,74 @@
+# frozen_string_literal: true
+
+module RuboCop
+  module Cop
+    module Style
+      # This cop checks the usage of pre-2.4 `Hash[args]` method of converting arrays to hashes
+      #
+      # @example
+      #   # bad
+      #   Hash[ary]
+      #
+      #   # good
+      #   ary.to_h
+      #
+      #   # bad
+      #   Hash[key1, value1, key2, value2]
+      #
+      #   # good
+      #   {key1 => value1, key2 => value2}
+      class LegacyHash < Base
+        extend AutoCorrector
+
+        MSG_TO_H = 'Prefer ary.to_h to Hash[ary].'
+        MSG_LITERAL = 'Prefer literal hash to Hash[arg1, arg2, ...].'
+        MSG_SPLAT = 'Prefer array_of_pairs.to_h to Hash[*array].'
```

Comment:
These messages should be have dynamic variable names. Please refer to other cops like [this one][1]

[1]: https://github.com/rubocop-hq/rubocop/blob/77db90e6fa0f45ea7cb0b6ec4bee23faf66a385a/lib/rubocop/cop/style/eval_with_location.rb#L42-L45

Reply:
I contemplated this, but I prefer not to. The real code that caused offense might look like this (Rubocop's own codebase):
```ruby
Hash[PERCENT_LITERAL_TYPES.map do |type|
  [type, preferred_delimiters_config[type] ||
     preferred_delimiters_config['default']]
end]
```
...and that's not the limit, block inside `Hash[]` can take many lines, so the suggestion would look really cumbersome. That's why I've decided to stick with abstract recommendations.

Reply:
@zverok that's a fair point. I think maybe one case we can be more explicit about is `Hash[]` to `{}`.

Reply:
@dvandersluis I actually believe it is the tiniest percent of actual `Hash::[]` usages in code, and didn't want to complicate the code for the rare case.

Reply:
Fair enough, I was basing on your example cases.